src/terminal_box.rs

```diff
@@ -406,72 +406,69 @@ where
+                                            }
+                                            index = i + 1;
+                                            pos += width;
+                                            break;
+                                        }
+                                    }
+                                    index
+                                };
+                                while pos < full_width {
+                                    let (width, x) = rects[index % rects.len()];
+                                    if let Some(height) = x {
+                                        let pos_offset = mk_pos_offset!(pos, height);
+                                        let underline_quad =
+                                            mk_quad!(pos_offset, style_line_height, width);
+                                        renderer.fill_quad(underline_quad, line_color);
+                                    }
+                                    pos += width;
+                                    index += 1;
                                 }
+                            };
+
+                            if metadata.flags.contains(Flags::DOTTED_UNDERLINE) {
+                                let bottom_offset = style_line_height * 2.0;
+                                draw_repeated(&[(2.0, Some(bottom_offset)), (3.0, None)]);
                             }
 
                             if metadata.flags.contains(Flags::DASHED_UNDERLINE) {
                                 let bottom_offset = style_line_height * 2.0;
-
-                                let full_width = self.end_x - self.start_x;
-                                let mut accu_width = 0.0;
-                                let mut dash_width = 6.0f32.min(full_width - accu_width);
-                                let gap_width = dash_width / 2.0;
-
-                                // gap-width dash first
-                                let pos_offset = mk_pos_offset!(accu_width, bottom_offset);
-                                let underline_quad =
-                                    mk_quad!(pos_offset, style_line_height, gap_width);
-                                renderer.fill_quad(underline_quad, line_color);
-                                accu_width += gap_width * 2.0;
-
-                                while accu_width < full_width {
-                                    dash_width = dash_width.min(full_width - accu_width);
-                                    let pos_offset = mk_pos_offset!(accu_width, bottom_offset);
-                                    let underline_quad =
-                                        mk_quad!(pos_offset, style_line_height, dash_width);
-                                    renderer.fill_quad(underline_quad, line_color);
-                                    accu_width += dash_width + gap_width;
-                                }
+                                draw_repeated(&[(6.0, Some(bottom_offset)), (3.0, None)]);
```

Comment:
Add a comment describing the tuples passed and/or bring back the descriptive variable names (`dash`, `gap`).